src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,408 @@
+{
+	auto n_samples = y.vlen;
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	float64_t grad_bias = 0;
+	switch (distribution)
+	{
+	case POISSON:
+		for (int i = 0; i < grad_mu.vlen; i++)
+		{
+			grad_bias += grad_mu[i];
+			grad_bias -= y[i] * grad_mu[i] / mu[i];
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	grad_bias /= n_samples;
+
+	return grad_bias;
+}
+
+bool GLM::train_machine(std::shared_ptr<Features> data = NULL)
+{
+	std::shared_ptr<SGDMinimizer> minimizer;
+
+	std::shared_ptr<GLMCostFunction> cost_function;
+	cost_function->set_target(std::make_shared<GLM> (this));
+	minimizer->set_cost_function(std::shared_ptr<FirstOrderCostFunction>);
+
+	std::shared_ptr<GradientDescendUpdater> gradient_updater;
+	minimizer->set_gradient_updater(gradient_updater);
+
+	minimizer->set_number_passes(10000);
+
+	std::shared_ptr<ConstLearningRate> learning_rate;
+	learning_rate->set_const_learning_rate(m_learning_rate);
+	minimizer->set_learning_rate(learning_rate);
+	
+	
+	minimizer->set_penalty_weight(m_lambda);
+
+	std::shared_ptr<ElasticNetPenalty> penalty;
+	penalty->set_l1_ratio(m_alpha);
+	minimizer->set_penalty_type(penalty);
+
+	minimizer->minimize();//returns min cost
+}
+
+
+
+
+class GLMCostFunction: public FirstOrderStochasticCostFunction
+{
+public:
+	GLMCostFunction():FirstOrderStochasticCostFunction() {  init(); }
+	virtual ~GLMCostFunction() {  }
+
+	void set_target(const std::shared_ptr<GLM>&obj)
+	{
+		require(obj, "Obj must set");
+		if(m_obj != obj)
+			m_obj=obj;
+	}
+
+	void unset_target()
+	{
+		m_obj=NULL;
+	}
+
+	virtual float64_t get_cost()
+	{
+		//TODO
+	}
+
+	virtual SGVector<float64_t> obtain_variable_reference()
+	{
+		require(m_obj,"Object not set");
+		m_derivatives = SGVector<float64_t>((m_obj->m_w).vlen);
+		return m_obj->m_w;
+	}
+
+	virtual SGVector<float64_t> get_gradient()
+	{
+		SGMatrix<float64_t> X = m_obj->LinearMachine::features->get_computed_dot_feature_matrix();
+		SGVector<float64_t> y = m_obj->m_labels->get_values();
+		SGVector<float64_t> w_old(m_obj->m_w);
+
+		auto n_samples = y.vlen;
+		auto n_features = X.num_rows;
+	
+		SGVector<float64_t> z = compute_z(X, m_obj->m_w, m_obj->bias);
+		SGVector<float64_t> mu = non_linearity(z);
+		SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+		SGVector<float64_t> grad_w(m_obj->m_w.vlen);
+		SGVector<float64_t> a;
+		switch (m_obj->distribution)
+		{
+		case POISSON:
+			a = y * grad_mu / mu;
+			linalg::transpose_matrix(linalg::add(linalg::matrix_prod(SGMatrix(grad_mu), X, true, false)), (linalg::matrix_prod(SGMatrix(a), X, true, false)), 1, -1);
+			break;
+	
+		default:
+			error("Not a valid distribution type");
+			break;
+		}
+		for (int i = 0; i < grad_w.vlen; i++)
+			grad_w[i] /= n_samples;
+	
+		return grad_w;
+	}
+
+	virtual void begin_sample()
```

Comment:
What is this function supposed to do? I am confused.
It is called in the `minimize()` function defined in the `SGDMinimizer` class.

Reply:
check how it is used in the various optimizers, might have to do with more advanced optimization methods
@yorkerlin might be able to help?

Reply:
I couldn't find the `FirstOrderStochasticCostFunction` class being inherited in any other class which is why I was unable to figure out. This method, `begin_sample()` belongs to the `FirstOrderStochasticCostFunction` class

Reply:
For random cost functions there might be some initialization needed. And sample just samples the cost function